cirq/linalg/transformations.py

```diff
@@ -304,3 +307,109 @@ def partial_trace(tensor: np.ndarray,
+        atol: The tolerance for determining that a factored state is pure.
+
+    Returns:
+        A mixture representation of the partial trace of the wavefunction.
+    """
+
+    tot_dims = 1 << len(wavefunction.shape)
+    keep_dims = 1 << len(keep_indices)
+    try:
+        # Attempt to do efficient state factoring.
+        state = keep_qubits(wavefunction, keep_indices).reshape(keep_dims, 1)
+        return ((1.0, state))
+    except ValueError:
+        # Fall back to a (non-unique) mixture representation.
+        rho = np.kron(np.conj(wavefunction.reshape(tot_dims, 1)).T,
+                      wavefunction.reshape(tot_dims, 1)).reshape((2, 2) * len(wavefunction.shape))
+        keep_rho = partial_trace(rho, keep_indices).reshape((keep_dims,) * 2)
+        w, v = np.linalg.eig(keep_rho)
+        mixture = tuple(zip(w, [vec.reshape((2,) * len(keep_indices)) for vec in v.T]))
+        return tuple(filter(lambda p: not approx_eq(p[0], 0.0), mixture))
+
+
+def keep_qubits(wavefunction: np.ndarray,
+                keep_indices: List[int],
+                atol: Union[int, float] = 1e-8) -> np.ndarray:
+    """Return a representation of the given state over a subset of its qubits.
+
+    The input wavfunction must have shape `(2,2,2,...)`.
+
+    A given wavefunction can be efficiently expressed over a subset of its
+    qubits by comparing magnitudes of unnormalized wavefunctions that remain
+    after each possible state over the qubits to be removed is sliced out of
+    the total input wavefunction. The input state will be factorizable into
+     `keep_indices` if the sum of the absolute value of inner products between
+    the renormalized largest-magnitude candidate and all candidates ("coherence
+    measure") is 1, indicating the output state is a pure state. Global phase
+    of the input state will not necessariy be preserved in the output state.
+
+    If the provided wavefunction cannot be factored into a pure state over
+    `keep_indices`, the method will fail and raise `ValueError`.
+
+    Args:
+        wavefunction: A wavefunction to express over a qubit subset.
+        keep_indices: Which indices to express the wavefunction on.
+        atol: The minimum tolerance for comparing the output state's coherence
+            measure to 1.
+
+    Returns:
+        The wavefunction expressed over the desired subset of qubits.
+
+    Raises:
+        ValueError: if the wavefunction is not of the correct shape or the
+        indices are not a valid subset of the input wavefunction's indices, or
+        the result of factoring is not a pure state.
+    """
+
+    if not np.isclose(np.linalg.norm(wavefunction), 1):
+        raise ValueError("Input state must be normalized.")
+    if not len(set(keep_indices)) == len(keep_indices):
+        raise ValueError(
+            "keep_indices were {} but must be unique.".format(keep_indices))
+    n_qubits = len(wavefunction.shape)
+    if n_qubits < len(keep_indices) or \
```

Comment:
Don't need the `n_qubits < len` check. It's implied by the distinctness + in range checks.

Reply:
good catch